src/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/clusterclient/SensorClientFragment.kt

```diff
@@ -0,0 +1,207 @@
+  }
+
+  private fun watchSensorButtonChecked() {
+    sensorWatchTimer = Timer("SensorWatchTimer", false).apply {
+      schedule(REFRESH_PERIOD_MS, REFRESH_PERIOD_MS) {
+        try {
+          readSensorCluster(clusterNameSpinner.selectedItem.toString(), true)
+        } catch (ex: Exception) {
+          showMessage(R.string.sensor_client_read_error_text, ex.toString())
+        }
+        // TODO: remove
+        consumeSensorValue((Calendar.getInstance().time.time.toLong() % 20).toDouble(), "%", true)
+      }
+    }
+  }
+
+  private fun watchSensorButtonUnchecked() {
+    sensorWatchTimer?.cancel()
+    sensorWatchTimer = null
+  }
+
+  private fun readSensorCluster(clusterName: String, addToGraph: Boolean) {
+    val deviceId = deviceIdEd.text.toString().toULong().toLong()
+    val endpointId = endpointIdEd.text.toString().toInt()
+    val clusterConfig = CLUSTERS[clusterName]
+    val clusterRead = clusterConfig!!["read"] as (Long, Int, ReadCallback) -> Unit
+
+    clusterRead(deviceId, endpointId, object : ReadCallback {
+      override fun onSuccess(value: Int) {
+        val unitValue = clusterConfig!!["unitValue"] as Double
+        val unitSymbol = clusterConfig!!["unitSymbol"] as String
+        consumeSensorValue(value * unitValue, unitSymbol, addToGraph)
+      }
+
+      override fun onError(ex: Exception) {
+        showMessage(R.string.sensor_client_read_error_text, ex.toString())
+      }
+    })
+  }
+
+  private fun consumeSensorValue(value: Double, unitSymbol: String, addToGraph: Boolean) {
+    requireActivity().runOnUiThread {
+      lastValueTv.text = requireContext().getString(
+          R.string.sensor_client_last_value_text, value, unitSymbol
+      )
+
+      if (addToGraph) {
+        // Make the graph visible on the first sample
+        if (sensorData.isEmpty)
+          sensorGraph.visibility = View.VISIBLE
+        val dataPoint = DataPoint(Calendar.getInstance().time, value)
+        sensorData.appendData(dataPoint, false, MAX_DATA_POINTS)
+      }
+    }
+  }
+
+  private fun showMessage(msgResId: Int, stringArgs: String? = null) {
+    requireActivity().runOnUiThread {
+      val context = requireContext()
+      val message = context.getString(msgResId, stringArgs)
+      Toast.makeText(context, message, Toast.LENGTH_SHORT).show()
+      Log.i(TAG, message)
+    }
+  }
+
+  companion object {
+    private const val TAG = "SensorClientFragment"
+    private const val REFRESH_PERIOD_MS = 2000L
+    private const val MAX_DATA_POINTS = 20
+
+    private val CLUSTERS = mapOf(
+        "Temperature" to mapOf(
+            "read" to { deviceId: Long, endpointId: Int, callback: ReadCallback ->
+              val device = ChipClient.getDeviceController().getDevicePointer(deviceId)
+              val cluster = ChipClusters.TemperatureMeasurementCluster(device, endpointId)
+              cluster.readMeasuredValueAttribute(callback)
+            },
+            "unitValue" to 0.01,
+            "unitSymbol" to "*C"
```

Comment:
I'm just curious, isn't it possible to print degree character ° by something like "\370"?

Reply:
I'll try.